app/graphql/api/v2/context.rb

```diff
@@ -10,6 +10,32 @@ def has_fragment?(name)
     end
   end
 
+  def authorized_demarche?(demarche)
+    if self[:authorized] == true
+      return true
+    end
+
+    # We are caching authorization logic because it is called for each node
+    # of the requested graph and can be expensive. Context is reset per request so it is safe.
+    self[:authorized] ||= {}
+    if self[:authorized][demarche.id]
+      return true
+    end
+
+    administrateur = demarche.administrateurs.find do |administrateur|
+      if self[:token]
+        administrateur.valid_api_token?(self[:token])
+      else
+        administrateur.id == self[:administrateur_id]
+      end
+    end
+
+    if administrateur
+      self[:authorized][demarche.id] = true
+      true
+    end
```

Comment:
J'ai l'impression de globablement comprendre cette logique, mais ça me fait un peu bizarre que par exemple on ne renvoie pas toujours explicitement une valeur, ou qu'on ne remplisse pas systématiquement le cache avec la valeur qu'on a trouvé. Je sais pas si c'est un comportement voulu ou pas.

Reply:
J'ai cherché un équivalent de ActiveSupport::Cache#fetch sur des Hash simple, mais j'ai pas trouvé.

Une solution pour rendre la logique explicite (et surtout s'assurer qu'une valeur est renvoyée dans tous les cas) serait de faire un lazy-hash :

```
def authorized_demarche?(demarche)
  if self[:authorized] == true
    return true
  end

  # We are caching authorization logic because it is called for each node
  # of the requested graph and can be expensive. Context is reset per request so it is safe.
  self[:authorized] ||= Hash.new do |hash, demarche_id|
    # Compute the hash value dynamically when first requested
    authorized_administrateur = demarche.administrateurs.find do |administrateur|
      if self[:token]
        administrateur.valid_api_token?(self[:token])
      else
        administrateur.id == self[:administrateur_id]
      end
    end
    hash[demarche_id] == authorized_administrateur.present?
  end

  self[:authorized][demarche.id]
end
```

Avec ça, la valeur est calculée la première fois qu'elle est demandée, et récupérée depuis le cache sinon. Tu en penses quoi ?

Reply:
Ça me paraît bien :)